### ISSUE TITLE
Heap Corruption due to PyObject.Dispose

### ISSUE THREAD

**Issue by PuerNoctis** [SciSharp/Keras.NET#94]
In one of our projects we encountered a consistent heap corruption during execution, when calling the `Predict` and `Fit` methods of our model. We experienced this behavior using the functional API as well as the `Sequential` API while running on .NET Core 3.1 and 2.0.

We then decided to debug into Keras and its dependencies and found out, that at some point a `PyObject` is being disposed, which has already been finalized by the Python.NET runtime. Within the `Dispose` method of the `PyObject`, it is not checked for finalization before decreasing the reference count and deallocating the pointer.

We opened a issue over at pythonnet for this (https://github.com/pythonnet/pythonnet/issues/1045)

At this point we can't gauge whether the heap corruptions we experienced are solely due to this problem, or whether Keras is disposing objects when it shouldn't. For now, we are running with a custom Python.NET build with the fix described applied.

If you need some code to reproduce the problem, we might be able to cut down everything unnecessary from our code and post it here.

**Comment by PuerNoctis** [comment]
I have analyzed the issue more thoroughly now, and the corruption seems to take place when the `kwargs` (or one of its contained objects) from the `InvokeMethod`-method are deallocated:

```csharp
public PyObject InvokeMethod(string method, Dictionary<string, object> args)
{
   var pyargs = ToTuple(new object[]
   {
        args.FirstOrDefault().Value
   });

    var kwargs = new PyDict();

    bool skip = true;
    foreach (var item in args)
    {
        if (skip)
        {
            skip = false;
            continue;
        }

        if (item.Value != null && !string.IsNullOrWhiteSpace(item.Value.ToString()))
        {
            kwargs[item.Key] = ToPython(item.Value);
        }
    }

    if (args.Count > 0)
        return PyInstance.InvokeMethod(method, pyargs, kwargs);
    else
        return PyInstance.InvokeMethod(method, null, null);
}
```
When `kwargs` leaves scope, its class finalizer will be called and pythonnet will call `Dispose` on that `PyDict`. The native resources for that object have already been freed however, which caused the heap corruption. Additionally, pythonnet reports invalid reference counts on the `PyDict` or the objects inside. I tried disposing `kwargs` manually after the `PyInstance.InvokeMethod` call with the same result.

This happens mostly during the `Fit` and sometimes during the `Predict` calls, but not right away, we have about 170 iterations until things start to crash.

Somewhere downstream it seems that the native python object is already deallocated or cleaned up, yet the `PyObject` still has references to it. 

It might be worth noting that we are running on Windows and using Tensorflow as backend, as well as 'libopenblas'.

More infos are over at the mentioned pythonnet issue. In the meanwhile I try to get a MCVE so I can give you a more reproducible code.
